lib/spack/spack/solver/concretize.lp

```diff
@@ -461,13 +461,58 @@ error(100, "Attempted to use external for '{0}' which does not satisfy any confi
 % Config required semantics
 %-----------------------------------------------------------------------------
 
-activate_requirement_rules(Package) :- attr("node", Package).
-activate_requirement_rules(Package) :- attr("virtual_node", Package).
+activate_requirement(Package, X) :-
+  attr("node", Package),
+  requirement_group(Package, X),
+  not requirement_conditional(Package, X).
+
+activate_requirement(Package, X) :-
+  attr("virtual_node", Package),
+  requirement_group(Package, X),
+  not requirement_conditional(Package, X).
+
+activate_requirement(Package, X) :-
+  attr("node", Package),
+  1 { condition_holds(Y) : requirement_group_member(Y, Package, X) },
+  requirement_group(Package, X),
+  requirement_conditional(Package, X).
+
+activate_requirement(Package, X) :-
+  attr("virtual_node", Package),
+  1 { condition_holds(Y) : requirement_group_member(Y, Package, X) },
+  requirement_group(Package, X),
+  requirement_conditional(Package, X).
+
+{ activate_requirement_rule(Y, Package, X) : requirement_group_member(Y, Package, X) } :-
+  activate_requirement(Package, X).
+
+impose(Y) :- activate_requirement_rule(Y, Package, X).
+do_not_impose(Y) :-
+  not activate_requirement_rule(Y, Package, X),
+  requirement_group_member(Y, Package, X).
+
+
+error(2, "cannot impose all the requirements for package '{0}'", Package)
+  :-  do_not_impose(Y),
+      condition_holds(Y),
+      not requirement_conditional(Package, X),
+      requirement_group_member(Y, Package, X).
+
+% Avoid impose(Y) to be derived by other rules when it is part of a requirement
```

Comment:
I'm not sure what this means and none of the `concretize_requirements.py` tests fail (including the new ones you added) when I remove this block.

Reply:
This is an integrity constraint and expresses a property of the solution that shouldn't be violated. In this case, it should never happen that `impose(Y)` is in the answer set when `activate_requirement_rule(Y, Package, X)` is not.